File: lib/services/billing_state.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../models/company_settings.dart';

/// Единая state machine billing (C3).
///
/// Stored statuses в Firestore: `active`, `trial`, `grace`, `suspended`, `cancelled`.
/// `blocked` — legacy alias для `suspended` (не писать в новые документы).
/// `trial_expired` / `past_due` — **не** stored; только UI/audit labels.
enum BillingDisplayPhase { active, trial, grace, blocked }

class BillingEvaluation {
  final bool allowsAccess;
  final BillingDisplayPhase displayPhase;
  final String storedStatus;
  final String? blockReason;
  final DateTime? trialUntil;
  final DateTime? graceUntil;

  const BillingEvaluation({
    required this.allowsAccess,
    required this.displayPhase,
    required this.storedStatus,
    this.blockReason,
    this.trialUntil,
    this.graceUntil,
  });

  /// UI key для blocked screens (`suspended`, `cancelled`, `payment_required`).
  String get blockUiKey {
    if (storedStatus == 'cancelled') return 'cancelled';
    if (storedStatus == 'suspended' || storedStatus == 'blocked') {
      return 'suspended';
    }
    if (blockReason == 'grace_expired' && storedStatus == 'trial') {
      return 'trial_expired';
    }
    return 'payment_required';
  }

  int graceDaysRemaining(DateTime now) {
    if (graceUntil == null) return 0;
    final d = graceUntil!.difference(now).inDays;
    return d < 0 ? 0 : d;
  }
}

class BillingState {
  BillingState._();

  static const canonicalStatuses = [
    'active',
    'trial',
    'grace',
    'suspended',
    'cancelled',
  ];

  /// Legacy alias — treat as suspended.
  static const legacyBlockedStatus = 'blocked';

  static const defaultGraceDays = 7;

  static BillingEvaluation evaluateFromSettings(
    CompanySettings settings, {
    DateTime? now,
  }) {
    return evaluateFromMap(
      {
        'billingStatus': settings.billingStatus,
        'trialUntil': settings.trialEndsAt,
        'trialEndsAt': settings.trialEndsAt,
        'paidUntil': settings.paidUntil,
        'gracePeriodDays': settings.gracePeriodDays,
      },
      now: now,
    );
  }

  static BillingEvaluation evaluateFromMap(
    Map<String, dynamic>? data, {
    DateTime? now,
  }) {
    final clock = now ?? DateTime.now();
    if (data == null) {
      return const BillingEvaluation(
        allowsAccess: false,
        displayPhase: BillingDisplayPhase.blocked,
        storedStatus: '',
        blockReason: 'missing_data',
      );
    }

    final status = data['billingStatus'] as String?;
    if (status == null || status.isEmpty) {
      return const BillingEvaluation(
        allowsAccess: false,
        displayPhase: BillingDisplayPhase.blocked,
        storedStatus: '',
        blockReason: 'missing_status',
      );
    }

    if (status == legacyBlockedStatus ||
        status == 'suspended' ||
        status == 'cancelled') {
      return BillingEvaluation(
        allowsAccess: false,
        displayPhase: BillingDisplayPhase.blocked,
        storedStatus: status,
        blockReason: status,
      );
    }

    if (status == 'active') {
      return BillingEvaluation(
        allowsAccess: true,
        displayPhase: BillingDisplayPhase.active,
        storedStatus: status,
      );
    }

    final graceDays = _graceDays(data);

    if (status == 'grace') {
      final anchor = _dateFrom(data['paidUntil']) ?? _dateFrom(data['trialUntil']);
      if (anchor == null) {
        return BillingEvaluation(
          allowsAccess: false,
          displayPhase: BillingDisplayPhase.blocked,
          storedStatus: status,
          blockReason: 'missing_paidUntil',
        );
      }
      final until = anchor.add(Duration(days: graceDays));
      if (clock.isBefore(until)) {
        return BillingEvaluation(
          allowsAccess: true,
          displayPhase: BillingDisplayPhase.grace,
          storedStatus: status,
          graceUntil: until,
        );
      }
      return BillingEvaluation(
        allowsAccess: false,
        displayPhase: BillingDisplayPhase.blocked,
        storedStatus: status,
        blockReason: 'grace_expired',
        graceUntil: until,
      );
    }

    if (status == 'trial') {
      final trialUntil =
          _dateFrom(data['trialUntil']) ?? _dateFrom(data['trialEndsAt']);
      if (trialUntil == null) {
        return BillingEvaluation(
          allowsAccess: false,
          displayPhase: BillingDisplayPhase.blocked,
          storedStatus: status,
          blockReason: 'missing_trialUntil',
        );
      }
      if (clock.isBefore(trialUntil)) {
        return BillingEvaluation(
          allowsAccess: true,
          displayPhase: BillingDisplayPhase.trial,
          storedStatus: status,
          trialUntil: trialUntil,
        );
      }
      final until = trialUntil.add(Duration(days: graceDays));
      if (clock.isBefore(until)) {
        return BillingEvaluation(
          allowsAccess: true,
          displayPhase: BillingDisplayPhase.grace,
          storedStatus: status,
          trialUntil: trialUntil,
          graceUntil: until,
        );
      }
      return BillingEvaluation(
        allowsAccess: false,
        displayPhase: BillingDisplayPhase.blocked,
        storedStatus: status,
        blockReason: 'grace_expired',
        trialUntil: trialUntil,
        graceUntil: until,
      );
    }

    return BillingEvaluation(
      allowsAccess: false,
      displayPhase: BillingDisplayPhase.blocked,
      storedStatus: status,
      blockReason: 'unknown_status',
    );
  }

  static int _graceDays(Map<String, dynamic> data) {
    final raw = data['gracePeriodDays'];
    if (raw is num && raw.toInt() > 0) return raw.toInt();
    return defaultGraceDays;
  }

  static DateTime? _dateFrom(dynamic v) {
    if (v is Timestamp) return v.toDate();
    if (v is DateTime) return v;
    return null;
  }
}

File: functions/lib/billingState.js
const DEFAULT_GRACE_DAYS = 7;

const CANONICAL_STATUSES = [
  "active",
  "trial",
  "grace",
  "suspended",
  "cancelled",
];

/** Legacy alias — treat as suspended. Not written to new docs. */
const LEGACY_BLOCKED_STATUS = "blocked";

function toDate(v) {
  if (!v) return null;
  if (v instanceof Date) return v;
  if (typeof v.toDate === "function") return v.toDate();
  return null;
}

function gracePeriodDays(data) {
  const d = data && data.gracePeriodDays;
  return typeof d === "number" && d > 0 ? d : DEFAULT_GRACE_DAYS;
}

function addDays(date, days) {
  const out = new Date(date.getTime());
  out.setDate(out.getDate() + days);
  return out;
}

function graceUntil(data, anchor) {
  if (!anchor) return null;
  return addDays(anchor, gracePeriodDays(data));
}

/**
 * Unified billing evaluation (C3).
 * Expired trial → grace window until trialUntil + gracePeriodDays.
 * Expired grace → deny until billingEnforcer sets suspended.
 */
function evaluateBilling(data, now = new Date()) {
  if (!data || typeof data !== "object") {
    return {
      allowsAccess: false,
      displayPhase: "blocked",
      storedStatus: "",
      blockReason: "missing_data",
    };
  }

  const status = data.billingStatus;
  if (!status || typeof status !== "string") {
    return {
      allowsAccess: false,
      displayPhase: "blocked",
      storedStatus: "",
      blockReason: "missing_status",
    };
  }

  if (
    status === LEGACY_BLOCKED_STATUS ||
    status === "suspended" ||
    status === "cancelled"
  ) {
    return {
      allowsAccess: false,
      displayPhase: "blocked",
      storedStatus: status,
      blockReason: status,
    };
  }

  if (status === "active") {
    return {
      allowsAccess: true,
      displayPhase: "active",
      storedStatus: status,
    };
  }

  if (status === "grace") {
    const anchor = toDate(data.paidUntil) || toDate(data.trialUntil);
    if (!anchor) {
      return {
        allowsAccess: false,
        displayPhase: "blocked",
        storedStatus: status,
        blockReason: "missing_paidUntil",
      };
    }
    const until = graceUntil(data, anchor);
    if (now < until) {
      return {
        allowsAccess: true,
        displayPhase: "grace",
        storedStatus: status,
        graceUntil: until,
      };
    }
    return {
      allowsAccess: false,
      displayPhase: "blocked",
      storedStatus: status,
      blockReason: "grace_expired",
      graceUntil: until,
    };
  }

  if (status === "trial") {
    const trialUntil = toDate(data.trialUntil) || toDate(data.trialEndsAt);
    if (!trialUntil) {
      return {
        allowsAccess: false,
        displayPhase: "blocked",
        storedStatus: status,
        blockReason: "missing_trialUntil",
      };
    }
    if (now < trialUntil) {
      return {
        allowsAccess: true,
        displayPhase: "trial",
        storedStatus: status,
        trialUntil,
      };
    }
    const until = graceUntil(data, trialUntil);
    if (now < until) {
      return {
        allowsAccess: true,
        displayPhase: "grace",
        storedStatus: status,
        trialUntil,
        graceUntil: until,
      };
    }
    return {
      allowsAccess: false,
      displayPhase: "blocked",
      storedStatus: status,
      blockReason: "grace_expired",
      trialUntil,
      graceUntil: until,
    };
  }

  return {
    allowsAccess: false,
    displayPhase: "blocked",
    storedStatus: status,
    blockReason: "unknown_status",
  };
}

function billingAllowsAccess(data, now = new Date()) {
  return evaluateBilling(data, now).allowsAccess;
}

module.exports = {
  DEFAULT_GRACE_DAYS,
  CANONICAL_STATUSES,
  LEGACY_BLOCKED_STATUS,
  evaluateBilling,
  billingAllowsAccess,
  graceUntil,
  gracePeriodDays,
};
